### PR TITLE
AzureAdHelper & AzureAdOptions

### PR DESCRIPTION
**Pull request by terencefan**

### CODE REVIEW

File: src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+        public string ClientSecret { get; private set; }
+
+        public string TenantId { get; private set; }
+
+        public bool Enabled { get; private set; } = false;
+
+        public string Instance { get; set; } = "https://login.microsoftonline.com/{0}";
+
+        public static AzureActiveDirectoryOptions Disabled = new AzureActiveDirectoryOptions();
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs:21]: order

Comment (vwxyzh) [reply src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs:21]: readonly?

File: src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+using System.Security.Cryptography.X509Certificates;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.Azure.SignalR
+{
+    public class AzureActiveDirectoryOptions
+    {
+        public readonly X509Certificate2 ClientCert;
+
+        public string ClientId { get; private set; }
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs:11]: `{get;}`

File: src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+        private static readonly Regex GuidRegex = new Regex("^[0-9a-z]{8}-[0-9a-z]{4}-[0-9a-z]{4}-[0-9a-z]{4}-[0-9a-z]{12}$");
+
+        private AzureActiveDirectoryOptions()
+        {
+        }
+
+        private AzureActiveDirectoryOptions(string tenantId)
+        {
+            var m = GuidRegex.Match(tenantId);
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs:31]: IsMatch

File: src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+        public string TenantId { get; private set; }
+
+        public bool Enabled { get; private set; } = false;
+
+        public string Instance { get; set; } = "https://login.microsoftonline.com/{0}";
+
+        public static AzureActiveDirectoryOptions Disabled = new AzureActiveDirectoryOptions();
+
+        private static readonly Regex GuidRegex = new Regex("^[0-9a-z]{8}-[0-9a-z]{4}-[0-9a-z]{4}-[0-9a-z]{4}-[0-9a-z]{12}$");
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs:23]: should we allow uppercase?

File: src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+        public string ClientSecret { get; }
+
+        public string TenantId { get; }
+
+        public bool Enabled { get; } = false;
+
+        public string Instance { get; set; } = "https://login.microsoftonline.com/{0}";
+
+        public static readonly AzureActiveDirectoryOptions Disabled = new AzureActiveDirectoryOptions();
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs:21]: order

File: src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+        public string ClientId { get; }
+
+        public string ClientSecret { get; }
+
+        public string TenantId { get; }
+
+        public bool Enabled { get; } = false;
+
+        public string Instance { get; set; } = "https://login.microsoftonline.com/{0}";
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs:19]: format?

File: src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs [hunk 7]
```diff
@@ -0,0 +1,65 @@
+﻿using System;
+using System.Security.Cryptography.X509Certificates;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.Azure.SignalR
+{
+    public class AzureActiveDirectoryOptions
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs:7]: >public [](start = 4, length = 6)

no `public` for now

File: src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+        private static readonly Regex GuidRegex = new Regex("^[0-9A-Za-z]{8}-[0-9A-Za-z]{4}-[0-9A-Za-z]{4}-[0-9A-Za-z]{4}-[0-9A-Za-z]{12}$");
+
+        private AzureActiveDirectoryOptions()
+        {
+        }
+
+        private AzureActiveDirectoryOptions(string tenantId)
+        {
+            if (!GuidRegex.IsMatch(tenantId))
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs:31]: is `Guid.TryParse` enough?

File: src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        public Uri BuildAuthority()
+        {
+            var s = string.Format(Instance, TenantId);
+            try
+            {
+                return new Uri(s);
+            }
+            catch (Exception)
+            {
+                throw new FormatException($"{Instance} is invalid, please refer to [link] for available Instances.");
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs:47]: move these validations to the `Validate` method of `ServiceOptionsExtensions`?

File: src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+            if (!GuidRegex.IsMatch(tenantId))
+            {
+                throw new FormatException($"{tenantId} is not a valid tenandId, should be [xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx]");
+            }
+            TenantId = tenantId;
+        }
+
+        public Uri BuildAuthority()
+        {
+            var s = string.Format(Instance, TenantId);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs:40]: `string.Format` can throw, also move this into `try..catch` block

File: src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs [hunk 11]
```diff
@@ -0,0 +1,65 @@
+﻿using System;
+using System.Security.Cryptography.X509Certificates;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.Azure.SignalR
+{
+    public class AzureActiveDirectoryOptions
+    {
+        public readonly X509Certificate2 ClientCert;
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs:9]: `public` means it is exposed to user, is it expected? or `internal`? and same questions apply to below properties.

File: src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs (hunk 6 above)
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs:19]: What does `Instance` stands for? better naming?
Also please add triple-slash comments to public properties

File: src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        public AzureActiveDirectoryOptions(string clientId, string clientSecret, string tenantId) : this(tenantId)
+        {
+            ClientId = clientId;
+            ClientSecret = clientSecret;
+            Enabled = true;
+        }
+
+        public AzureActiveDirectoryOptions(string clientId, X509Certificate2 clientCert, string tenantId) : this(tenantId)
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs:58]: looks like `this(tenantId)` can be `this(clientId, tenantId)` and `Enabled = true;` can be moved into it?

File: src/Microsoft.Azure.SignalR.Common/Utilities/HttpClientFactory.cs [hunk 13]
```diff
@@ -8,21 +8,17 @@ namespace Microsoft.Azure.SignalR
...
-            ClientFactory = services.GetRequiredService<IHttpClientFactory>();
-        }
-
-        public static HttpClient CreateClient()
-        {
-            return ClientFactory.CreateClient();
+            _httpClientFactory = services.GetRequiredService<IHttpClientFactory>();
         }
 
+        public static HttpClient Client => _httpClientFactory.CreateClient();
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Utilities/HttpClientFactory.cs:22]: I think the previous `CreateClient()` is better, the semantic of property sounds more like some value that `get` costs not expensive and rarely change

File: src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs (hunk 11 above)
Comment (terencefan) [reply src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs:9]: I think there is no reason that we don't allow our users to read a property set by themselves?

File: src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs [hunk 14]
```diff
@@ -0,0 +1,93 @@
+﻿using System;
+using System.Security.Cryptography.X509Certificates;
+
+namespace Microsoft.Azure.SignalR
+{
+    public class AzureActiveDirectoryOptions
+    {
+        internal static string GlobalFormat = "https://login.microsoftonline.com/{0}";
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs:8]: `const`

File: src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+        internal static string DogfoodFormat = "https://login.windows-ppe.net/{0}";
+        internal static string USGovernmentFormat = "https://login.microsoftonline.us/{0}";
+        internal static string GermanyFormat = "https://login.microsoftonline.de/{0}";
+        internal static string ChinaFormat = "https://login.chinacloudapi.cn/{0}";
+
+        private string _format = GlobalFormat;
+
+        public static readonly AzureActiveDirectoryOptions Disabled = new AzureActiveDirectoryOptions();
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs:16]: `public static` to upper most

File: src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+
+        public readonly X509Certificate2 ClientCert;
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs:18]: use property instead

File: src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+
+        public string ClientId { get; }
+
+        public string ClientSecret { get; }
+
+        public bool Enabled { get; } = false;
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs:24]: ~~= false~~ the default value is `false`, no need to set

File: src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+            return this;
+        }
+
+        public AzureActiveDirectoryOptions WithChina()
+        {
+            _format = ChinaFormat;
+            return this;
+        }
+
+        private AzureActiveDirectoryOptions(string clientId, string tenantId)
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs:72]: move to where other ctor sits

File: src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+        {
+            if (!Guid.TryParseExact(tenantId, "D", out _)) {
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Auth/AzureActiveDirectoryOptions.cs:74]: `{` to another line

File: test/Microsoft.Azure.SignalR.Common.Tests/AzureActiveDirectoryTests.cs [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+
+            var app = AzureActiveDirectoryHelper.BuildApplication(options);
+
+            var token = await app.AcquireTokenForClient(AzureActiveDirectoryOptions.DefaultScopes).WithSendX5C(true).ExecuteAsync();
+
+            ConfigurationManager<OpenIdConnectConfiguration> configManager = new ConfigurationManager<OpenIdConnectConfiguration>(
+                options.BuildMetadataAddress(),
+                new OpenIdConnectConfigurationRetriever()
+            );
+            var keys = configManager.GetConfigurationAsync().Result.SigningKeys;
```
Comment (vwxyzh) [inline test/Microsoft.Azure.SignalR.Common.Tests/AzureActiveDirectoryTests.cs:33]: await

**Pull request merged**